packages-exp/auth-exp/src/core/user/token_manager.ts

```diff
@@ -99,12 +99,34 @@ export class StsTokenManager {
     return null;
   }
 
-  // TODO(lisajian): Fill in once refresh listener is added
   private async getTokenAndTriggerCallback(
-    _auth: AuthInternal,
-    _forceRefresh = false
+    auth: AuthInternal,
+    forceRefresh = false
   ): Promise<string | null> {
-    return null;
+    if (forceRefresh) {
+      // TODO(lisajian): Add and use client error code - ERROR_TOKEN_REFRESH_UNAVAILABLE
+      _assert(
+        auth._refreshWithCustomTokenProvider,
+        auth,
+        AuthErrorCode.INTERNAL_ERROR
+      );
+      return auth._refreshWithCustomTokenProvider();
+    }
+
+    if (!this.isExpired) {
+      return this.accessToken;
+    }
+
+    // isExpired includes a buffer window, during which the access token will still be returned even
+    // if a customTokenProvider is not set yet.
+    _assert(
+      this.expirationTime && Date.now() < this.expirationTime,
+      auth,
+      AuthErrorCode.TOKEN_EXPIRED
+    );
+    return auth._refreshWithCustomTokenProvider
+      ? auth._refreshWithCustomTokenProvider()
```

Comment:
Does there need to be an `_assert` here too to check if `_refreshWithCustomTokenProvider` is set when the access token is expired?

Reply:
Ah, so this is the case when we're in a buffer window (token is not yet expired but will expire in `Buffer.TOKEN_REFRESH`) - if `_refreshWithCustomTokenProvider` is not set, we just return `this.accessToken`